<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>EvoTraceR â€¢ EvoTraceR</title>
<script src="deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet">
<link href="deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet">
<script src="deps/headroom-0.11.0/headroom.min.js"></script><script src="deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="deps/search-1.0.0/fuse.min.js"></script><script src="deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="pkgdown.js"></script><meta property="og:title" content="EvoTraceR">
<meta name="description" content="Package to identifiy and analyse Amplicon Sequence Variants after performing an experiment through the EVoBC kit.">
<meta property="og:description" content="Package to identifiy and analyse Amplicon Sequence Variants after performing an experiment through the EVoBC kit.">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="index.html">EvoTraceR</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.0.3</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="articles/1_introduction.html">Introduction</a></li>
    <li><a class="dropdown-item" href="articles/2_running_REvoBC.html">Running EvoTraceR</a></li>
  </ul>
</li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/Nowak-Lab/EvoTraceR/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-home">
<div class="row">
  <main id="main" class="col-md-9"><div class="section level1">
<div class="page-header"><h1 id="evotracer---evolutionary-lineage-tracing-in-r">EvoTraceR - Evolutionary lineage Tracing in R<a class="anchor" aria-label="anchor" href="#evotracer---evolutionary-lineage-tracing-in-r"></a>
</h1></div>
<p><a href="https://github.com/Nowak-Lab/EvoTraceR/actions?query=workflow%3Acheck-master" class="external-link"><img src="https://github.com/Nowak-Lab/EvoTraceR/workflows/check-master/badge.svg" alt="Actions Status"></a> <a href="https://github.com/Nowak-Lab/EvoTraceR/actions?query=workflow%3Acheck-development" class="external-link"><img src="https://github.com/Nowak-Lab/EvoTraceR/workflows/check-development/badge.svg" alt="Actions Status"></a></p>
<p><em>EvoTraceR</em> is an R package to analyse sequencing amplicon data from CRISPR-Cas9 recorder lineage tracing experiments. The package takes in paired-end FASTQ files from one to many tissues. The sequenced amplicon can contain one to many Cas9 cut sites. <em>EvoTraceR</em> trims and merges reads, collapses duplicates, calls mutations and infers a tree. The package outputs the inferred tree of relationships between Amplicon Sequence Variants (ASVs) as well as summary plots and tables of mutations.</p>
<div class="float">
<img src="articles/figures/evotracer_pipeline_concept.png" alt="EvoTraceR pipeline concept"><div class="figcaption">EvoTraceR pipeline concept</div>
</div>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<p>The package can be installed using <code>devtools</code>.</p>
<pre><code><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://devtools.r-lib.org/" class="external-link">devtools</a></span><span class="op">)</span></span>
<span><span class="fu">install_github</span><span class="op">(</span><span class="st">"Nowak-Lab/EvoTraceR"</span><span class="op">)</span></span></code></pre>
<p><em>EvoTraceR</em> can also be installed with the above method in a <code>conda</code> environment. Based on our experience, a clean environment installed with <code>r-base</code>, <code>r-essentials</code>, and <code>r-devtools</code> from the <code>conda-forge</code> channel works well.</p>
<pre><code>conda create -n evotracer
conda install -c conda-forge r-base
conda install -c conda-forge r-essentials
conda install -c conda-forge r-devtools</code></pre>
</div>
<div class="section level2">
<h2 id="dependencies">Dependencies<a class="anchor" aria-label="anchor" href="#dependencies"></a>
</h2>
<ul>
<li><a href="https://ccb.jhu.edu/software/FLASH/" class="external-link">flash</a></li>
<li><a href="http://www.usadellab.org/cms/?page=trimmomatic" class="external-link">trimmomatic</a></li>
</ul>
<p>Binaries for these dependencies can be downloaded via the links above. All other dependencies are handled by <code>devtools</code> during the package installation.</p>
</div>
<div class="section level2">
<h2 id="usage-example">Usage example<a class="anchor" aria-label="anchor" href="#usage-example"></a>
</h2>
<p>A simulated toy dataset is available to test <em>EvoTraceR</em>. The sequences were generated from a single simulated mouse <code>MMUS1469</code> and taken from cancer populations in two tissues (labelled <code>PRL</code> and <code>HMR</code>). The CRISPR-Cas9 recorder design was designated <code>BC10v0</code>. The naming scheme for the FASTQ files was thus <code>MMUS1469_PRL_BC10v0_MG_120419</code>. Note that <em>EvoTraceR</em> expects similar input file names with tissue labels in the same position of the naming scheme.</p>
<p>To get started with the analysis, we provide the directory containing the input FASTQ files. We also provide an output directory and the paths to the <code>trimmomatic</code> and <code>flash</code> binaries on our system (adjust these paths as needed for your system). Then we preprocess our input data with the <code>initialize_EvoTraceR</code> function.</p>
<pre><code><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/Nowak-Lab/EvoTraceR" class="external-link">EvoTraceR</a></span><span class="op">)</span></span>
<span><span class="va">input_dir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"input"</span>, package <span class="op">=</span> <span class="st">"EvoTraceR"</span><span class="op">)</span></span>
<span><span class="va">output_dir</span> <span class="op">&lt;-</span> <span class="st">"example_output"</span></span>
<span><span class="va">trimmomatic_path</span> <span class="op">&lt;-</span> <span class="st">"/my/path/trimmomatic.jar"</span></span>
<span><span class="va">flash_path</span> <span class="op">&lt;-</span> <span class="st">"/my/path/flash"</span></span>
<span></span>
<span><span class="va">EvoTraceR_object</span> <span class="op">&lt;-</span></span>
<span>  <span class="fu"><a href="reference/initialize_EvoTraceR.html">initialize_EvoTraceR</a></span><span class="op">(</span></span>
<span>    input_dir <span class="op">=</span> <span class="va">input_dir</span>,</span>
<span>    output_dir <span class="op">=</span> <span class="va">output_dir</span>,</span>
<span>    trimmomatic_path <span class="op">=</span> <span class="va">trimmomatic_path</span>,</span>
<span>    flash_path <span class="op">=</span> <span class="va">flash_path</span><span class="op">)</span></span></code></pre>
<p>Now that the FASTQ data has been cleaned and the paired reads merged, we align the merged reads to the reference and call indels. To do this, we use the <code>asv_analysis</code> function and provide our unedited template sequence as the <code>ref_seq</code> and the expected flanking sequences. We also provide the 1-based reference sequence coordinates of the expected Cas9 cut sites and the borders between individual target regions (e.g., guide=20bp, PAM=3bp, spacer=3bp). This information is used to filter sequences with unexpected mutation patterns. For additional options on alignment parameters, parallelization and more, see the documentation.</p>
<pre><code><span><span class="va">EvoTraceR_object</span> <span class="op">&lt;-</span></span>
<span>  <span class="fu"><a href="reference/asv_analysis.html">asv_analysis</a></span><span class="op">(</span>EvoTraceR_object <span class="op">=</span> <span class="va">EvoTraceR_object</span>,</span>
<span>               ref_name <span class="op">=</span> <span class="st">"BC10v0"</span>,</span>
<span>               ref_seq <span class="op">=</span> <span class="st">"TCTACACGCGCGTTCAACCGAGGAAAACTACACACACGTTCAACCACGGTTTTTTACACACGCATTCAACCACGGACTGCTACACACGCACTCAACCGTGGATATTTACATACTCGTTCAACCGTGGATTGTTACACCCGCGTTCAACCAGGGTCAGATACACCCACGTTCAACCGTGGTACTATACTCGGGCATTCAACCGCGGCTTTCTGCACACGCCTACAACCGCGGAACTATACACGTGCATTCACCCGTGGATC"</span>,</span>
<span>               ref_flank_left <span class="op">=</span> <span class="st">"^TCTAC"</span>,</span>
<span>               ref_flank_right <span class="op">=</span> <span class="st">"CCCGTGGATC$"</span>,</span>
<span>               ref_cut_sites <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">17</span>, <span class="fl">43</span>, <span class="fl">69</span>, <span class="fl">95</span>, <span class="fl">121</span>, <span class="fl">147</span>, <span class="fl">173</span>, <span class="fl">199</span>, <span class="fl">225</span>, <span class="fl">251</span><span class="op">)</span>,</span>
<span>               ref_border_sites <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">26</span>, <span class="fl">52</span>, <span class="fl">78</span>, <span class="fl">104</span>, <span class="fl">130</span>, <span class="fl">156</span>, <span class="fl">182</span>, <span class="fl">208</span>, <span class="fl">234</span><span class="op">)</span><span class="op">)</span></span></code></pre>
<p>Finally, we process the mutations, infer a tree and output summaries of the results.</p>
<pre><code><span><span class="va">EvoTraceR_object</span> <span class="op">&lt;-</span></span>
<span>  <span class="fu"><a href="reference/analyse_mutations.html">analyse_mutations</a></span><span class="op">(</span>EvoTraceR_object <span class="op">=</span> <span class="va">EvoTraceR_object</span><span class="op">)</span></span>
<span></span>
<span><span class="va">EvoTraceR_object</span> <span class="op">&lt;-</span></span>
<span>  <span class="fu"><a href="reference/infer_phylogeny.html">infer_phylogeny</a></span><span class="op">(</span>EvoTraceR_object <span class="op">=</span> <span class="va">EvoTraceR_object</span>, mutations_use <span class="op">=</span> <span class="st">"del_ins"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">EvoTraceR_object</span> <span class="op">&lt;-</span></span>
<span>  <span class="fu"><a href="reference/create_df_summary.html">create_df_summary</a></span><span class="op">(</span><span class="va">EvoTraceR_object</span><span class="op">)</span></span></code></pre>
</div>
<div class="section level2">
<h2 id="output-files">Output files<a class="anchor" aria-label="anchor" href="#output-files"></a>
</h2>
<p>After running the example analysis, the following output files will be generated.</p>
<pre><code>example_output/
|-- asv_analysis
|   |-- asv_filtering_freq.pdf
|   `-- asv_length_freq.pdf
|-- fastq_analysis
|   |-- fastq_flash_merged
|   |-- fastq_summary.csv
|   `-- fastq_trimmed
`-- phylogeny_analysis
    `-- phylogeny_del_ins
        |-- asv_stat.csv
        |-- tree_all_clones.csv
        `-- tree_all_clones.newick</code></pre>
<p>The two key files for most follow-up analyses are as follows.</p>
<ul>
<li>
<code>asv_stat.csv</code>: Summary of all of detected amplicon sequence variants (ASVs) and mutations in the input FASTQ files</li>
<li>
<code>tree_all_clones.newick</code>: Tree of the phylogenetic relationships between ASVs based on the detected mutations</li>
</ul>
</div>
<div class="section level2">
<h2 id="contact">Contact<a class="anchor" aria-label="anchor" href="#contact"></a>
</h2>
<p>Please feel free to contact us with feedback: Dawid Nowak, dgn2001 at med.cornell.edu or Armin Scheben, ascheben at cshl.edu.</p>
</div>
</div>
  </main><aside class="col-md-3"><div class="links">
<h2 data-toc-skip>Links</h2>
<ul class="list-unstyled">
<li><a href="https://github.com/Nowak-Lab/EvoTraceR/" class="external-link">Browse source code</a></li>
<li><a href="https://github.com/Nowak-Lab/EvoTraceR" class="external-link">Report a bug</a></li>
</ul>
</div>

<div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li>file <a href="LICENSE-text.html">LICENSE</a>
</li>
</ul>
</div>


<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing EvoTraceR</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>Lucrezia Patruno <br><small class="roles"> Maintainer, author </small>  </li>
<li>Daniele Ramazzotti <br><small class="roles"> Author </small> <a href="https://orcid.org/0000-0002-6087-2666" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
<li>Armin Scheben <br><small class="roles"> Author </small> <a href="https://orcid.org/0000-0002-2230-2013" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
<li>Stephen Staklinski <br><small class="roles"> Author </small>  </li>
<li>Rebecca Hassett <br><small class="roles"> Author </small>  </li>
<li>Dylan Riffle <br><small class="roles"> Author </small>  </li>
<li>Dawid Nowak <br><small class="roles"> Author </small>  </li>
</ul>
</div>



  </aside>
</div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Lucrezia Patruno, Daniele Ramazzotti, Armin Scheben, Stephen Staklinski, Rebecca Hassett, Dylan Riffle, Dawid Nowak.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.9000.</p>
</div>

    </footer>
</div>





  </body>
</html>
