<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>EvoTraceR â€¢ EvoTraceR</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="bootstrap-toc.css">
<script src="bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script><meta property="og:title" content="EvoTraceR">
<meta property="og:description" content="Package to identifiy and analyse Amplicon Sequence Variants after performing an experiment through the EVoBC kit.">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-home">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">EvoTraceR</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.99.27</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="articles/1_introduction.html">Introduction</a>
    </li>
    <li>
      <a href="articles/2_running_REvoBC.html">Running EvoTraceR</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/Nowak-Lab/EvoTraceR/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">
<div class="section level1">
<div class="page-header"><h1 id="evotracer">EvoTraceR<a class="anchor" aria-label="anchor" href="#evotracer"></a>
</h1></div>
<p><a href="https://github.com/Nowak-Lab/EvoTraceR/actions?query=workflow%3Acheck-master" class="external-link"><img src="https://github.com/Nowak-Lab/EvoTraceR/workflows/check-master/badge.svg" alt="Actions Status"></a> <a href="https://github.com/Nowak-Lab/EvoTraceR/actions?query=workflow%3Acheck-development" class="external-link"><img src="https://github.com/Nowak-Lab/EvoTraceR/workflows/check-development/badge.svg" alt="Actions Status"></a></p>
<p><em>EvoTraceR</em> is an R package to analyse sequencing amplicon data from CRISPR-Cas9 recorder lineage tracing experiments. The package takes in paired-end FASTQ files from one to many tissues. The sequenced amplicon can contain one to many Cas9 cut sites. <em>EvoTraceR</em> trims and merges reads, collapses duplicates, calls mutations and infers a tree. The package outputs the inferred tree of relationships between Amplicon Sequence Variants (ASVs) as well as summary plots and tables of mutations.</p>
<div class="figure">
<img src="articles/figures/evotracer_pipeline_concept.png" alt=""><p class="caption">EvoTraceR pipeline concept</p>
</div>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<pre><code><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://devtools.r-lib.org/" class="external-link">devtools</a></span><span class="op">)</span></span>
<span><span class="fu">install_github</span><span class="op">(</span><span class="st">"Nowak-Lab/EvoTraceR"</span><span class="op">)</span></span></code></pre>
<p><em>EvoTraceR</em> can also be installed with the above method in a <code>conda</code> environment. Based on our experience, a clean environment installed with <code>r-base</code> and <code>r-essentials</code> from the <code>conda-forge</code> channel works well.</p>
</div>
<div class="section level2">
<h2 id="dependencies">Dependencies<a class="anchor" aria-label="anchor" href="#dependencies"></a>
</h2>
<ul>
<li><a href="https://ccb.jhu.edu/software/FLASH/" class="external-link">flash</a></li>
<li><a href="http://www.usadellab.org/cms/?page=trimmomatic" class="external-link">trimmomatic</a></li>
</ul>
<p>Binaries for these dependencies can be downloaded via the links above. All other dependencies are handled by <code>devtools</code> during the package installation.</p>
</div>
<div class="section level2">
<h2 id="usage-example">Usage example<a class="anchor" aria-label="anchor" href="#usage-example"></a>
</h2>
<p>An empirical toy dataset is available to test <em>EvoTraceR</em>. The sequences were generated from a single mouse <code>MMUS1495</code> and taken from cancer populations in four tissues (labelled <code>PRL</code>,<code>RBL</code>,<code>LVR</code>,<code>LVM</code>). The CRISPR-Cas9 recorder design was designated <code>BC10v0</code>. The naming scheme for the FASTQ files was thus <code>MMUS1469_PRL_BC10v0_MG_120419</code>. Note that <em>EvoTraceR</em> expects similar input file names with tissue labels in the same position of the naming scheme.</p>
<p>To get started with the analysis, we provide the directory containing the input FASTQ files. We also provide an output directory and the paths to the <code>trimmomatic</code> and <code>flash</code> binaries on our system (adjust these paths as needed for your system). Then we preprocess our input data with the <code>initialize_EvoTraceR</code> function.</p>
<pre><code><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/Nowak-Lab/EvoTraceR" class="external-link">EvoTraceR</a></span><span class="op">)</span></span>
<span><span class="va">input_dir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"input"</span>, package <span class="op">=</span> <span class="st">"EvoTraceR"</span><span class="op">)</span></span>
<span><span class="va">output_dir</span> <span class="op">&lt;-</span> <span class="st">"example_output"</span></span>
<span><span class="va">trimmomatic_path</span> <span class="op">&lt;-</span> <span class="st">"/my/path/trimmomatic.jar"</span></span>
<span><span class="va">flash_path</span> <span class="op">&lt;-</span> <span class="st">"/my/path/flash"</span></span>
<span></span>
<span><span class="va">EvoTraceR_object</span> <span class="op">&lt;-</span></span>
<span>  <span class="fu"><a href="reference/initialize_EvoTraceR.html">initialize_EvoTraceR</a></span><span class="op">(</span></span>
<span>    input_dir <span class="op">=</span> <span class="va">input_dir</span>,</span>
<span>    output_dir <span class="op">=</span> <span class="va">output_dir</span>,</span>
<span>    map_file_sample <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>    trimmomatic_path <span class="op">=</span> <span class="va">trimmomatic_path</span>,</span>
<span>    flash_path <span class="op">=</span> <span class="va">flash_path</span><span class="op">)</span></span></code></pre>
<p>Now that the FASTQ data has been cleaned and the paired reads merged, we align the merged reads to the reference and call indels. To do this, we use the <code>asv_analysis</code> function and provide our unedited template sequence as the <code>ref_seq</code> and the expected flanking sequences. We also provide the 1-based reference sequence coordinates of the expected Cas9 cut sites and the borders between individual target regions. This information is used to filter sequences with unexpected mutation patterns. For additional options on alignment parameters, parallelization and more, see the documentation.</p>
<pre><code><span><span class="va">EvoTraceR_object</span> <span class="op">&lt;-</span></span>
<span>  <span class="fu"><a href="reference/asv_analysis.html">asv_analysis</a></span><span class="op">(</span>EvoTraceR_object <span class="op">=</span> <span class="va">EvoTraceR_object</span>,</span>
<span>               ref_name <span class="op">=</span> <span class="st">"BC10v0"</span>,</span>
<span>               ref_seq <span class="op">=</span> <span class="st">"TCTACACGCGCGTTCAACCGAGGAAAACTACACACACGTTCAACCACGGTTTTTTACACACGCATTCAACCACGGACTGCTACACACGCACTCAACCGTGGATATTTACATACTCGTTCAACCGTGGATTGTTACACCCGCGTTCAACCAGGGTCAGATACACCCACGTTCAACCGTGGTACTATACTCGGGCATTCAACCGCGGCTTTCTGCACACGCCTACAACCGCGGAACTATACACGTGCATTCACCCGTGGATC"</span>,</span>
<span>               ref_flank_left <span class="op">=</span> <span class="st">"^TCTAC"</span>,</span>
<span>               ref_flank_right <span class="op">=</span> <span class="st">"CCCGTGGATC$"</span>,</span>
<span>               ref_cut_sites <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">17</span>, <span class="fl">43</span>, <span class="fl">69</span>, <span class="fl">95</span>, <span class="fl">121</span>, <span class="fl">147</span>, <span class="fl">173</span>, <span class="fl">199</span>, <span class="fl">225</span>, <span class="fl">251</span><span class="op">)</span>, <span class="co"># cut sites for Cas9 based on "ref_seq"</span></span>
<span>               ref_border_sites <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">26</span>, <span class="fl">52</span>, <span class="fl">78</span>, <span class="fl">104</span>, <span class="fl">130</span>, <span class="fl">156</span>, <span class="fl">182</span>, <span class="fl">208</span>, <span class="fl">234</span><span class="op">)</span> <span class="co"># border for marking site (guide=20, PAM=3, spacer=3) = 26 x bp</span></span>
<span>               <span class="op">)</span></span></code></pre>
<p>Finally, we process the mutations, infer a tree and output summaries of the results.</p>
<pre><code><span><span class="va">EvoTraceR_object</span> <span class="op">&lt;-</span></span>
<span>  <span class="fu"><a href="reference/analyse_mutations.html">analyse_mutations</a></span><span class="op">(</span>EvoTraceR_object <span class="op">=</span> <span class="va">EvoTraceR_object</span><span class="op">)</span></span>
<span></span>
<span><span class="va">EvoTraceR_object</span> <span class="op">&lt;-</span></span>
<span>  <span class="fu"><a href="reference/infer_phylogeny.html">infer_phylogeny</a></span><span class="op">(</span>EvoTraceR_object <span class="op">=</span> <span class="va">EvoTraceR_object</span>, mutations_use <span class="op">=</span> <span class="st">"del_ins"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">EvoTraceR_object</span> <span class="op">&lt;-</span></span>
<span>  <span class="fu"><a href="reference/create_df_summary.html">create_df_summary</a></span><span class="op">(</span><span class="va">EvoTraceR_object</span><span class="op">)</span></span></code></pre>
</div>
<div class="section level2">
<h2 id="output-files">Output files<a class="anchor" aria-label="anchor" href="#output-files"></a>
</h2>
<p>After running the example analysis, the following output files will be generated.</p>
<pre><code>example_output/
|-- asv_analysis
|   |-- asv_filtering_freq.pdf
|   `-- asv_length_freq.pdf
|-- fastq_analysis
|   |-- fastq_flash_merged
|   |-- fastq_summary.csv
|   `-- fastq_trimmed
`-- phylogeny_analysis
    `-- phylogeny_del_ins
        |-- asv_stat.csv
        |-- tree_all_clones.csv
        `-- tree_all_clones.newick</code></pre>
<p>The two key files for most follow-up analyses are as follows.</p>
<ul>
<li>
<code>asv_stat.csv</code>: Summary of all of detected amplicon sequence variants (ASVs) and mutations in the input FASTQ files</li>
<li>
<code>tree_all_clones.newick</code>: Tree of the phylogenetic relationships between ASVs based on the detected mutations</li>
</ul>
</div>
<div class="section level2">
<h2 id="contact">Contact<a class="anchor" aria-label="anchor" href="#contact"></a>
</h2>
<p>Please feel free to contact us with feedback: Dawid Nowak, dgn2001 at med.cornell.edu or Armin Scheben, ascheben at cshl.edu.</p>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <div class="links">
<h2 data-toc-skip>Links</h2>
<ul class="list-unstyled">
<li><a href="https://github.com/Nowak-Lab/EvoTraceR/" class="external-link">Browse source code</a></li>
<li><a href="https://github.com/Nowak-Lab/EvoTraceR" class="external-link">Report a bug</a></li>
</ul>
</div>

<div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li>file <a href="LICENSE-text.html">LICENSE</a>
</li>
</ul>
</div>


<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing EvoTraceR</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>Lucrezia Patruno <br><small class="roles"> Maintainer, author </small>  </li>
<li>Daniele Ramazzotti <br><small class="roles"> Author </small> <a href="https://orcid.org/0000-0002-6087-2666" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
<li>Dawid Nowak <br><small class="roles"> Author </small>  </li>
</ul>
</div>



  </div>
</div>


      <footer><div class="copyright">
  <p></p>
<p>Developed by Lucrezia Patruno, Daniele Ramazzotti, Dawid Nowak.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.9000.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
