<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>initialize_REvoBC — initialize_REvoBC • REvoBC</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="initialize_REvoBC — initialize_REvoBC" />
<meta property="og:description" content="This function initializes the REvoBC object, by computing the set of Amplicon Sequence Variants.
It wraps all the steps performed by dada2, that are the following:
(1) Filter and trimming reads.
(2) Learning the error rates.
(3) Dereplication.
(4) Sample Inference.
(5) Merging forward and reverse reads.
(6) Construction of the Amplicon Sequence Variant table (ASV). 
(7) Removal of chimeras.
(8) [Optional] Production of a summary which tracks changes in the number of reads at every step of the pipeline." />






<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  

  

  </head>

  <body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">REvoBC</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.99.5</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/1_introduction.html">Introduction</a>
    </li>
    <li>
      <a href="../articles/2_running_REvoBC.html">Running REvoBC</a>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/Nowak-Lab/REvoBC/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>initialize_REvoBC</h1>
    <small class="dont-index">Source: <a href='https://github.com/Nowak-Lab/REvoBC/blob/master/R/user_functions.R'><code>R/user_functions.R</code></a></small>
    <div class="hidden name"><code>initialize_REvoBC.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>This function initializes the REvoBC object, by computing the set of Amplicon Sequence Variants.
It wraps all the steps performed by dada2, that are the following:
(1) Filter and trimming reads.
(2) Learning the error rates.
(3) Dereplication.
(4) Sample Inference.
(5) Merging forward and reverse reads.
(6) Construction of the Amplicon Sequence Variant table (ASV). 
(7) Removal of chimeras.
(8) [Optional] Production of a summary which tracks changes in the number of reads at every step of the pipeline.</p>
    </div>

    <div class="ref-usage sourceCode"><pre class='sourceCode r'><code><span class='fu'>initialize_REvoBC</span><span class='op'>(</span>
  <span class='va'>output_dir</span>,
  input_dir <span class='op'>=</span> <span class='cn'>NULL</span>,
  dada2_output_sequences <span class='op'>=</span> <span class='cn'>NULL</span>,
  output_dir_dada2 <span class='op'>=</span> <span class='cn'>NULL</span>,
  random_seed <span class='op'>=</span> <span class='cn'>NULL</span>,
  output_figures <span class='op'>=</span> <span class='cn'>TRUE</span>,
  multithread <span class='op'>=</span> <span class='cn'>TRUE</span>,
  map_file_sample <span class='op'>=</span> <span class='cn'>NULL</span>,
  dada2_pooled_analysis <span class='op'>=</span> <span class='cn'>FALSE</span>,
  dada2_chimeras_minFoldParentOverAbundance <span class='op'>=</span> <span class='fl'>8</span>,
  verbose <span class='op'>=</span> <span class='cn'>TRUE</span>,
  <span class='va'>...</span>
<span class='op'>)</span></code></pre></div>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>output_dir</th>
      <td><p>(Required). Path to the directory where all output files will be stored. The following <code>.csv</code> files will be created:</p><ul>
<li><p><code>quality_track_reads.csv</code>: track of the number of sequences during the 
different stepsof <code>dada2</code> analysis.</p></li>
<li><p><code> dada2_asv_prefilter.csv</code>: sequences detected by dada2, with the counts detected in each sample.</p></li>
</ul></td>
    </tr>
    <tr>
      <th>input_dir</th>
      <td><p>Path to the directory containing <code>.fastq</code> files for forward and reverse reads.
This folder should contain the fastq files (2 for each sample) with the following name pattern:
FILEPREFIX_SAMPLE_BARCODEVERSION_R1.fastq FILEPREFIX_SAMPLE_BARCODEVERSION_R1.fastq. SAMPLE refers to either an organ (in case multiple organs were sequenced)
or timepoint (if longitudinal data are provided). Note that REvoBC does not support mixed sample types (i.e. samples must be either all from organs or all from timepoints).
Required when <code>dada2_output_sequences</code> is null (i.e. when the user has not previously run dada2).</p></td>
    </tr>
    <tr>
      <th>dada2_output_sequences</th>
      <td><p>(Optional). In case users have already run dada2 up to the removal of bimeras, they should provide the path to the csv file containing the output (See description).
The output should contain samples on rows and sequences on columns.</p></td>
    </tr>
    <tr>
      <th>output_dir_dada2</th>
      <td><p>(Optional). Output folder for dada2 filtered fastqs. If it doesn't exist it is created. Default is NULL and results in creating a 
sub-folder in <code>output_folder</code> named 'filtered_fastq/'.</p></td>
    </tr>
    <tr>
      <th>random_seed</th>
      <td><p>(Optional). Seed for reproducibility. Deafult is NULL.</p></td>
    </tr>
    <tr>
      <th>output_figures</th>
      <td><p>(Optional). Boolean indicating whether or not to save intermediate dada2 plots such as read quality profiles.
Deafult is <code>TRUE</code> and corresponds to creating a sub-folder inside <code>output_dir</code> named "dada2_plots".
This folder will contain the following plots:</p><ul>
<li><p>Quality profiles of both forward and reverse reads (generated by <code>dada2</code>).</p></li>
<li><p>Error rates (generated by <code>dada2</code>).</p></li>
<li><p>Histogram of sequences length.</p></li>
</ul></td>
    </tr>
    <tr>
      <th>multithread</th>
      <td><p>(Optional) default <code>TRUE</code>. Whether to enable multithreading. If TRUE the number of threads is determined automatically (dada2 functionality).
If an integer is given, the number of threads is determined by its value.</p></td>
    </tr>
    <tr>
      <th>map_file_sample</th>
      <td><p>(Optional). In case fastq files names are not in the format <br /> FILEPREFIX_SAMPLE_FILESUFFIX_RX.fastq),
then users should provide a list that associates each filename (without the suffix _R1 and _R2) to the corresponding organ/day.
(e.g., if we have forward and reverse files named file1_R1.fastq and file1_R2.fastq that correspond to organ PRL (code for Prostate Left), than the parameter should be set as: <code>map_file_sample = c("file1" = "PRL")</code>).</p></td>
    </tr>
    <tr>
      <th>dada2_pooled_analysis</th>
      <td><p>(Optional). Deafault = FALSE. Boolean that is passed to dada2 function <code>dada</code>, which performs sample inference.
It can be set to TRUE in case of multiple samples coming from the same mouse (e.g. different organs or multiple time points) (See <a href='https://benjjneb.github.io/dada2/pool.html'>here</a> for more information).</p></td>
    </tr>
    <tr>
      <th>dada2_chimeras_minFoldParentOverAbundance</th>
      <td><p>(Optional). Deafult = 8. parameter passed to the <code>dada2</code> function <code>isBimeraDenovo</code> during the call to <code>dada2 removeBimeraDenovo</code>.</p></td>
    </tr>
    <tr>
      <th>verbose</th>
      <td><p>(Optional). Default TRUE. Boolean indicating whether or not to print the text output of dada2 functions.</p></td>
    </tr>
    <tr>
      <th>...</th>
      <td><p>(Optional) Any additional parameters passed to <code>dada2</code> functions.</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>An object of type REvoBC, which is a list that will contain the following fields:</p><ul>
<li><p><code>fastq_directory</code>: directory where the input fastq files are located.</p></li>
<li><p><code>output_directory</code>: directory where all the output files are being stored.</p></li>
<li><p><code>map_file_sample</code>: dataframe has as many rows as the input datasets, and for each input stores the sample (e.g. organ or day for longitudinal data)
to which it is associated.</p></li>
<li><p><code>dada2_asv_prefilter</code>: dataframe that stores all sequences detected by <code>dada2</code>. Note that
these sequences still need to be filtered (see also <code><a href='asv_analysis.html'>asv_analysis</a></code>).</p></li>
<li><p><code>dada2</code>: list which contains the percentage of chimeras found by <code>dada2</code> and a dataframe
that tracks the number of sequences during all <code>dada2</code> steps.</p></li>
</ul><p>. This function also saves the following <code>.csv files</code> in the subfolder <code>dada2_files</code> created in the output dirctory provided by the user:</p><ul>
<li><p>dada2_asv_prefilter.csv: sequences detected by dada2, with the counts detected in each sample.</p></li>
<li><p>quality_track_reads.csv: track of the number of sequences during all <code>dada2</code> steps.</p></li>
</ul>

    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p>For further information please refer to <a href='https://benjjneb.github.io/dada2/tutorial_1_8.html'>the original vignette</a>.
If users wish to manually run each step of dada2 computation they should execute all steps reported
in the vignette, up to the Removal of chimeras and save the output of <code>removeBimeraDenovo</code> in a csv file and provide the path to that file as parameter <code>dada2_output_sequences</code>.</p>

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <div class="ref-examples sourceCode"><pre class='sourceCode r'><code><span class='r-in'><span class='va'>input_dir</span> <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/system.file.html'>system.file</a></span><span class='op'>(</span><span class='st'>"extdata"</span>, <span class='st'>"input"</span>, package <span class='op'>=</span> <span class='st'>"REvoBC"</span><span class='op'>)</span></span>
<span class='r-in'><span class='va'>output_dir</span> <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/system.file.html'>system.file</a></span><span class='op'>(</span><span class='st'>"extdata"</span>, <span class='st'>"output"</span>, package <span class='op'>=</span> <span class='st'>"REvoBC"</span><span class='op'>)</span></span>
<span class='r-in'><span class='fu'>initialize_REvoBC</span><span class='op'>(</span>input_dir <span class='op'>=</span> <span class='va'>input_dir</span>, output_dir <span class='op'>=</span> <span class='va'>output_dir</span><span class='op'>)</span></span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> <span style='color: #00BBBB;'>ℹ</span> Found zipped fastq files, extracting</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> <span style='color: #00BBBB;'>ℹ</span> Done extracting</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span> Found 2 fastq files</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> <span style='color: #00BBBB;'>ℹ</span> The folliwng files, mapped to the corresponding sample, were found: </span>
<span class='r-out co'><span class='r-pr'>#&gt;</span> MMUS1469_HMR_BC10v0_MG_120419  :  HMR </span>
<span class='r-out co'><span class='r-pr'>#&gt;</span> Working directory:  /tmp/RtmpPkbDxu/filef1866eab5fb0/referencePath:  /home/runner/work/_temp/Library/REvoBC/extdata/input</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> <span style='color: #00BBBB;'>ℹ</span> Creating quality profiles figures</span>
<span class='r-wrn co'><span class='r-pr'>#&gt;</span> <span class='warning'>Warning: </span>`guides(&lt;scale&gt; = FALSE)` is deprecated. Please use `guides(&lt;scale&gt; = "none")` instead.</span>
<span class='r-wrn co'><span class='r-pr'>#&gt;</span> <span class='warning'>Warning: </span>`guides(&lt;scale&gt; = FALSE)` is deprecated. Please use `guides(&lt;scale&gt; = "none")` instead.</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> <span style='color: #00BBBB;'>ℹ</span> Filtering and trimming</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> Read in 59148 paired-sequences, output 54321 (91.8%) filtered paired-sequences.</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> <span style='color: #00BBBB;'>ℹ</span> Learning errors for forward reads</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span> 10908088 total bases in 54321 reads from 1 samples will be used for learning the error rates.</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> <span style='color: #00BBBB;'>ℹ</span> Learning errors for backward reads</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span> 11131107 total bases in 54321 reads from 1 samples will be used for learning the error rates.</span>
<span class='r-wrn co'><span class='r-pr'>#&gt;</span> <span class='warning'>Warning: </span>Transformation introduced infinite values in continuous y-axis</span>
<span class='r-wrn co'><span class='r-pr'>#&gt;</span> <span class='warning'>Warning: </span>Transformation introduced infinite values in continuous y-axis</span>
<span class='r-wrn co'><span class='r-pr'>#&gt;</span> <span class='warning'>Warning: </span>Transformation introduced infinite values in continuous y-axis</span>
<span class='r-wrn co'><span class='r-pr'>#&gt;</span> <span class='warning'>Warning: </span>Transformation introduced infinite values in continuous y-axis</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> <span style='color: #00BBBB;'>ℹ</span> Dereplicating fastqs</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> Dereplicating sequence entries in Fastq file: /home/runner/work/_temp/Library/REvoBC/extdata/output/filtered_fastq/MMUS1469_HMR_BC10v0_MG_120419_F_filt.fastq.gz</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> Encountered 12705 unique sequences from 54321 total sequences read.</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> Dereplicating sequence entries in Fastq file: /home/runner/work/_temp/Library/REvoBC/extdata/output/filtered_fastq/MMUS1469_HMR_BC10v0_MG_120419_R_filt.fastq.gz</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> Encountered 14618 unique sequences from 54321 total sequences read.</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> <span style='color: #00BBBB;'>ℹ</span> Inferring samples</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span> Sample 1 - 54321 reads in 12705 unique sequences.</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span> Sample 1 - 54321 reads in 14618 unique sequences.</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> 40387 paired-reads (in 170 unique pairings) successfully merged out of 51398 (in 693 pairings) input.</span>
<span class='r-wrn co'><span class='r-pr'>#&gt;</span> <span class='warning'>Warning: </span>Removed 2 rows containing missing values (geom_bar).</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> <span style='color: #00BBBB;'>ℹ</span> Removing bimeras</span>
<span class='r-msg co'><span class='r-pr'>#&gt;</span> Identified 59 bimeras out of 170 input sequences.</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span> $fastq_directory</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span> [1] "/home/runner/work/_temp/Library/REvoBC/extdata/input"</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span> </span>
<span class='r-out co'><span class='r-pr'>#&gt;</span> $output_directory</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span> [1] "/home/runner/work/_temp/Library/REvoBC/extdata/output"</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span> </span>
<span class='r-out co'><span class='r-pr'>#&gt;</span> $dada2</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span> $dada2$track</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span>     input filtered denoisedF denoisedR merged nonchim</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span> HMR 59148    54321     53760     51812  40387   38851</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span> </span>
<span class='r-out co'><span class='r-pr'>#&gt;</span> $dada2$bimera_percentage</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span> [1] 96.1968</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span> </span>
<span class='r-out co'><span class='r-pr'>#&gt;</span> $dada2$original_sequences</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span> [1] 170</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span> </span>
<span class='r-out co'><span class='r-pr'>#&gt;</span> </span>
<span class='r-out co'><span class='r-pr'>#&gt;</span> $map_file_sample</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span>                               sample</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span> MMUS1469_HMR_BC10v0_MG_120419    HMR</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span> </span>
<span class='r-out co'><span class='r-pr'>#&gt;</span> $dada2_asv_prefilter</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span> <span style='color: #949494;'># A tibble: 111 × 3</span></span>
<span class='r-out co'><span class='r-pr'>#&gt;</span>      HMR seq                                                           seq_names</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span>    <span style='color: #949494; font-style: italic;'>&lt;int&gt;</span> <span style='color: #949494; font-style: italic;'>&lt;chr&gt;</span>                                                         <span style='color: #949494; font-style: italic;'>&lt;chr&gt;</span>    </span>
<span class='r-out co'><span class='r-pr'>#&gt;</span> <span style='color: #BCBCBC;'> 1</span> <span style='text-decoration: underline;'>15</span>731 TCTACACGCGCGTTCAACCGAGGAAAACTACACACACGTTCAACCACGGTTTTTTACACA… SEQ001   </span>
<span class='r-out co'><span class='r-pr'>#&gt;</span> <span style='color: #BCBCBC;'> 2</span>  <span style='text-decoration: underline;'>2</span>669 TCTACACGCGCGTTCAACCGAGGAAAACTACACACGCATTCAACCACGGTTTATTACACG… SEQ002   </span>
<span class='r-out co'><span class='r-pr'>#&gt;</span> <span style='color: #BCBCBC;'> 3</span>  <span style='text-decoration: underline;'>2</span>308 GCTCCCTTCCCTTCCCTTCTCTCCGGCCCGCGGCGGCCGCGCTCCTCACCTGGCCAGGGG… SEQ003   </span>
<span class='r-out co'><span class='r-pr'>#&gt;</span> <span style='color: #BCBCBC;'> 4</span>  <span style='text-decoration: underline;'>2</span>157 TCTACACGCGCGTTCAACCGAGGAAAACTACACACACGTTCAACCACGGTTTGTTACACA… SEQ004   </span>
<span class='r-out co'><span class='r-pr'>#&gt;</span> <span style='color: #BCBCBC;'> 5</span>  <span style='text-decoration: underline;'>2</span>121 TCTACACGCGCGTTCAACCGAGGAAAACTACACACACATTCAACCGTGGTTTACTACACA… SEQ005   </span>
<span class='r-out co'><span class='r-pr'>#&gt;</span> <span style='color: #BCBCBC;'> 6</span>  <span style='text-decoration: underline;'>2</span>098 TCTACACGCGCGTTCAACCGAGGAAAACTACACACACATTCAACCGTGGTTTATTACACG… SEQ006   </span>
<span class='r-out co'><span class='r-pr'>#&gt;</span> <span style='color: #BCBCBC;'> 7</span>  <span style='text-decoration: underline;'>1</span>443 CTGGGCGGCAACAATCCCATTGCGTGTC                                  SEQ007   </span>
<span class='r-out co'><span class='r-pr'>#&gt;</span> <span style='color: #BCBCBC;'> 8</span>   896 GCTCCCTTCCCTTCCCTTCTCTCCGGCCCGCGGCGGCCGCGCTCCTCACCTGGCCAGGGG… SEQ008   </span>
<span class='r-out co'><span class='r-pr'>#&gt;</span> <span style='color: #BCBCBC;'> 9</span>   724 CTGGCACTCGTTGCTGGGTCTGGGGACGTC                                SEQ009   </span>
<span class='r-out co'><span class='r-pr'>#&gt;</span> <span style='color: #BCBCBC;'>10</span>   608 TCTACACGCGCGTTCAACTACACACACGTTCAACCACGGTTTTTTACACACGCATTCAAC… SEQ010   </span>
<span class='r-out co'><span class='r-pr'>#&gt;</span> <span style='color: #949494;'># … with 101 more rows</span></span>
<span class='r-out co'><span class='r-pr'>#&gt;</span> </span>
<span class='r-out co'><span class='r-pr'>#&gt;</span> attr(,"class")</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span> [1] "REvoBC"</span>
<span class='r-in'></span>
</code></pre></div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p><p>Developed by Lucrezia Patruno, Daniele Ramazzotti, Dawid Nowak.</p></p>
</div>

<div class="pkgdown">
  <p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 1.6.1.9001.</p></p>
</div>

      </footer>
   </div>

  


  

  </body>
</html>


