<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>perform_msa — perform_msa • REvoBC</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="perform_msa — perform_msa"><meta property="og:description" content="This function performs multi-sequence alignment and outputs statistics about mutations. 
Multiple Sequence Alignment is performed through the MUSCLE algorithm.
This function also compute the binary matrix indicating the presence/absence of
mutations in each ASV. Each mutation is characterized by a start position and an end position, it is thus identified through an ID
which indicates the start, end and type of mutation."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">REvoBC</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.99.10</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/1_introduction.html">Introduction</a>
    </li>
    <li>
      <a href="../articles/2_running_REvoBC.html">Running REvoBC</a>
    </li>
  </ul></li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/Nowak-Lab/REvoBC/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>perform_msa</h1>
    <small class="dont-index">Source: <a href="https://github.com/Nowak-Lab/REvoBC/blob/HEAD/R/user_functions.R" class="external-link"><code>R/user_functions.R</code></a></small>
    <div class="hidden name"><code>perform_msa.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>This function performs multi-sequence alignment and outputs statistics about mutations. 
Multiple Sequence Alignment is performed through the <a href="https://bioconductor.org/packages/release/bioc/html/muscle.html" class="external-link">MUSCLE</a> algorithm.
This function also compute the binary matrix indicating the presence/absence of
mutations in each ASV. Each mutation is characterized by a start position and an end position, it is thus identified through an ID
which indicates the start, end and type of mutation.</p>
    </div>

    <div class="ref-usage sourceCode"><pre class="sourceCode r"><code><span class="fu">perform_msa</span><span class="op">(</span><span class="va">REvoBC_object</span>, <span class="va">...</span><span class="op">)</span></code></pre></div>

    <div id="ref-sections">
    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>REvoBC_object</dt>
<dd><p>REvoBC object on which we want to perform msa.</p></dd>
<dt>...</dt>
<dd><p>Optional parameters (options and flags) passed to MUSCLE. See the <a href="http://www.drive5.com/muscle/muscle_userguide3.8.html" class="external-link">original guide</a> for detailed information on all possible options.</p></dd>
</dl></div>
    <div id="value">
    <h2>Value</h2>
    <p>REvoBC object with a new field named <code>alignment</code>, which is a list with the following fileds:</p><ul><li><p><code>msa_stringset</code>: output of MSA peformed with MUSCLE.</p></li>
<li><p><code>mutations_df</code> : (deprecated, will be removed).</p></li>
<li><p><code>asv_barcode_alignment</code>: tibble where each line corresponds to a position in a ASV, and the columns encode the following information:</p><ul><li><p>asv_names: name of the ASV</p></li>
<li><p>sample: sample identifier</p></li>
<li><p>position_bc260: position of the alteration in the original barcode. Note that insertions
are assigned to the position that coincides with their beginning.</p></li>
<li><p>alt: type of alteration. wt = Wild Type (i.e. non-mutated position). sub = substitution. del = deletion. ins = insertion.</p></li>
<li><p>ref_asv,</p></li>
<li><p>read_asv: respectively, the reference nucleotide observed in the original barcode and the one observed on the sequence.</p></li>
</ul></li>
<li><p><code>ASV_alterations_width</code>: number of alterations for each type in each ASV</p></li>
<li><p><code>mutations_coordinates</code>: tibble that stores all mutations identified on each ASV, indicating the start and end position and the nucleotides involved in the mutation</p></li>
<li><p><code>binary_mutation_matrix</code>: binary matrix encoding presence/absence of mutations on ASVs.</p></li>
<li><p><code>mutations_frequency</code>: tibble where for each position in each ASV you have the information about the frequency of the observed ASV in the sample.
This is useful to convey an idea about what is the fraction of total ASVs in each sample that are affected by that mutation in that position.</p></li>
</ul><p>. 
It also creates a new field in the REvoBC object called <code>smoothed deletions</code>, which contains the following information for
all the deletions identified in the ASVs, whose start and end site have been smoothed using the known cutting sites.
Smoothing is performed in the following way: the start site of each deletion is assigned to the cutting site on their left if that
is less distant then 5 nucleotides, otherwise they are assigned to the cutting site on their right. The end site is assigned to the 
cutting site on its right if that is less than 5 nucleotides distant, otherwise it is assigned to the cutting site
on its left.
The following dataframes are stored in the field <code>smoothed deletions</code>:</p><ul><li><p>binary_matrix: matrix |ASV| x |mutations|, which contains 1/0 if the mutation is respectively present or not in the ASV.</p></li>
<li><p>coordinate_matrix: matrix which stores all the information regardin the smoothed deletions. Information include start and end before smoothing and the deleted sequence.</p></li>
</ul><p>In addition, the following files are saved in the sub-folder "msa" created in the output directory chosen by the user:</p><ul><li><p>dnastringset.fa: fasta file where the sequences are stored.</p></li>
<li><p>dnastringset.fa: same as fasta, but in csv format.</p></li>
<li><p>dnastringset_muscle-muscle_msa.fasta: fasta with the stringset resulted from MUSCLE.</p></li>
<li><p>ASV_alterations_width.csv: number of alterations for each type in each ASV</p></li>
<li><p>mutations_frequency.csv: per sample normalized frequency of each alteration type, computed for each position of the barcode in each sample.</p></li>
<li><p>mutations_df.csv: content of mutations_df variable explained above.</p></li>
<li><p>mutations_coordinates: content of mutations_coordinates variable explained above.</p></li>
<li><p>binary_mutation_matrix: content of binary_mutation_matrix variable explained above.</p></li>
</ul><p>All identified mutations are displayed in a heatmap, saved inside the folder <code>msa_figures</code>.
This function also produces an output figure that contain the frequency of 
deletion, substitutions and insertions found in the different samples.</p>
    </div>
    </div>

    <div id="examples">
    <h2>Examples</h2>
    <div class="ref-examples sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">revo_analyzed</span><span class="op">)</span></span>
<span class="r-in"><span class="va">output_dir</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"output"</span>, package <span class="op">=</span> <span class="st">"REvoBC"</span><span class="op">)</span></span>
<span class="r-in"><span class="va">revo_analyzed</span><span class="op">$</span><span class="va">output_directory</span> <span class="op">=</span> <span class="va">output_dir</span></span>
<span class="r-in"><span class="va">revo_msa</span> <span class="op">=</span> <span class="fu">perform_msa</span><span class="op">(</span><span class="va">revo_analyzed</span><span class="op">)</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> MUSCLE v3.8.31 by Robert C. Edgar</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> http://www.drive5.com/muscle</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> This software is donated to the public domain.</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Please cite: Edgar, R.C. Nucleic Acids Res 32(5), 1792-97.</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> filefb366106b091 20 seqs, max length 261, avg  length 176</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1676 MB(24%)00:00:00                Iter   1    0.48%  K-mer dist pass 11676 MB(24%)00:00:00                Iter   1  100.00%  K-mer dist pass 1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1676 MB(24%)00:00:00                Iter   1    0.48%  K-mer dist pass 21676 MB(24%)00:00:00                Iter   1  100.00%  K-mer dist pass 2</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1676 MB(24%)00:00:00                Iter   1    5.26%  Align node       1676 MB(24%)00:00:00                Iter   1   10.53%  Align node1676 MB(24%)00:00:00                Iter   1   15.79%  Align node1676 MB(24%)00:00:00                Iter   1   21.05%  Align node1676 MB(24%)00:00:00                Iter   1   26.32%  Align node1676 MB(24%)00:00:00                Iter   1   31.58%  Align node1676 MB(24%)00:00:00                Iter   1   36.84%  Align node1676 MB(24%)00:00:00                Iter   1   42.11%  Align node1676 MB(24%)00:00:00                Iter   1   47.37%  Align node1676 MB(24%)00:00:00                Iter   1   52.63%  Align node1676 MB(24%)00:00:00                Iter   1   57.89%  Align node1676 MB(24%)00:00:00                Iter   1   63.16%  Align node1676 MB(24%)00:00:00                Iter   1   68.42%  Align node1676 MB(24%)00:00:00                Iter   1   73.68%  Align node1676 MB(24%)00:00:00                Iter   1   78.95%  Align node1676 MB(24%)00:00:00                Iter   1   84.21%  Align node1676 MB(24%)00:00:00                Iter   1   89.47%  Align node1676 MB(24%)00:00:00                Iter   1   94.74%  Align node1676 MB(24%)00:00:00                Iter   1  100.00%  Align node1676 MB(24%)00:00:00                Iter   1  100.00%  Align node</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1676 MB(24%)00:00:00                Iter   1    5.00%  Root alignment1676 MB(24%)00:00:00                Iter   1   10.00%  Root alignment1676 MB(24%)00:00:00                Iter   1   15.00%  Root alignment1676 MB(24%)00:00:00                Iter   1   20.00%  Root alignment1676 MB(24%)00:00:00                Iter   1   25.00%  Root alignment1676 MB(24%)00:00:00                Iter   1   30.00%  Root alignment1676 MB(24%)00:00:00                Iter   1   35.00%  Root alignment1676 MB(24%)00:00:00                Iter   1   40.00%  Root alignment1676 MB(24%)00:00:00                Iter   1   45.00%  Root alignment1676 MB(24%)00:00:00                Iter   1   50.00%  Root alignment1676 MB(24%)00:00:00                Iter   1   55.00%  Root alignment1676 MB(24%)00:00:00                Iter   1   60.00%  Root alignment1676 MB(24%)00:00:00                Iter   1   65.00%  Root alignment1676 MB(24%)00:00:00                Iter   1   70.00%  Root alignment1676 MB(24%)00:00:00                Iter   1   75.00%  Root alignment1676 MB(24%)00:00:00                Iter   1   80.00%  Root alignment1676 MB(24%)00:00:00                Iter   1   85.00%  Root alignment1676 MB(24%)00:00:00                Iter   1   90.00%  Root alignment1676 MB(24%)00:00:00                Iter   1   95.00%  Root alignment1676 MB(24%)00:00:00                Iter   1  100.00%  Root alignment1676 MB(24%)00:00:00                Iter   1  100.00%  Root alignment</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1676 MB(24%)00:00:00                Iter   2    5.56%  Refine tree   1676 MB(24%)00:00:00                Iter   2   11.11%  Refine tree1676 MB(24%)00:00:00                Iter   2   16.67%  Refine tree1676 MB(24%)00:00:00                Iter   2   22.22%  Refine tree1676 MB(24%)00:00:00                Iter   2   27.78%  Refine tree1676 MB(24%)00:00:00                Iter   2   33.33%  Refine tree1676 MB(24%)00:00:00                Iter   2   38.89%  Refine tree1676 MB(24%)00:00:00                Iter   2   44.44%  Refine tree1676 MB(24%)00:00:00                Iter   2   50.00%  Refine tree1676 MB(24%)00:00:00                Iter   2   55.56%  Refine tree1676 MB(24%)00:00:00                Iter   2   61.11%  Refine tree1676 MB(24%)00:00:00                Iter   2   66.67%  Refine tree1676 MB(24%)00:00:00                Iter   2   72.22%  Refine tree1676 MB(24%)00:00:00                Iter   2   77.78%  Refine tree1676 MB(24%)00:00:00                Iter   2   83.33%  Refine tree1676 MB(24%)00:00:00                Iter   2   88.89%  Refine tree1676 MB(24%)00:00:00                Iter   2   94.44%  Refine tree1676 MB(24%)00:00:00                Iter   2  100.00%  Refine tree1676 MB(24%)00:00:00                Iter   2  105.56%  Refine tree1676 MB(24%)00:00:00                Iter   2  100.00%  Refine tree</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1676 MB(24%)00:00:00                Iter   2    5.00%  Root alignment1676 MB(24%)00:00:00                Iter   2   10.00%  Root alignment1676 MB(24%)00:00:00                Iter   2   15.00%  Root alignment1676 MB(24%)00:00:00                Iter   2   20.00%  Root alignment1676 MB(24%)00:00:00                Iter   2   25.00%  Root alignment1676 MB(24%)00:00:00                Iter   2   30.00%  Root alignment1676 MB(24%)00:00:00                Iter   2   35.00%  Root alignment1676 MB(24%)00:00:00                Iter   2   40.00%  Root alignment1676 MB(24%)00:00:00                Iter   2   45.00%  Root alignment1676 MB(24%)00:00:00                Iter   2   50.00%  Root alignment1676 MB(24%)00:00:00                Iter   2   55.00%  Root alignment1676 MB(24%)00:00:00                Iter   2   60.00%  Root alignment1676 MB(24%)00:00:00                Iter   2   65.00%  Root alignment1676 MB(24%)00:00:00                Iter   2   70.00%  Root alignment1676 MB(24%)00:00:00                Iter   2   75.00%  Root alignment1676 MB(24%)00:00:00                Iter   2   80.00%  Root alignment1676 MB(24%)00:00:00                Iter   2   85.00%  Root alignment1676 MB(24%)00:00:00                Iter   2   90.00%  Root alignment1676 MB(24%)00:00:00                Iter   2   95.00%  Root alignment1676 MB(24%)00:00:00                Iter   2  100.00%  Root alignment1676 MB(24%)00:00:00                Iter   2  100.00%  Root alignment</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1676 MB(24%)00:00:00                Iter   2  100.00%  Root alignment</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1676 MB(24%)00:00:00                Iter   3    5.41%  Refine biparts1676 MB(24%)00:00:00                Iter   3    8.11%  Refine biparts1676 MB(24%)00:00:00                Iter   3   10.81%  Refine biparts1676 MB(24%)00:00:00                Iter   3   13.51%  Refine biparts1676 MB(24%)00:00:00                Iter   3   16.22%  Refine biparts1676 MB(24%)00:00:00                Iter   3   18.92%  Refine biparts1676 MB(24%)00:00:00                Iter   3   21.62%  Refine biparts1676 MB(24%)00:00:00                Iter   3   24.32%  Refine biparts1676 MB(24%)00:00:00                Iter   3   27.03%  Refine biparts1676 MB(24%)00:00:00                Iter   3   29.73%  Refine biparts1676 MB(24%)00:00:00                Iter   3   32.43%  Refine biparts1676 MB(24%)00:00:00                Iter   3   35.14%  Refine biparts1676 MB(24%)00:00:00                Iter   3   37.84%  Refine biparts1676 MB(24%)00:00:00                Iter   3   40.54%  Refine biparts1676 MB(24%)00:00:00                Iter   3   43.24%  Refine biparts1676 MB(24%)00:00:00                Iter   3   45.95%  Refine biparts1676 MB(24%)00:00:00                Iter   3   48.65%  Refine biparts1676 MB(24%)00:00:00                Iter   3   51.35%  Refine biparts1676 MB(24%)00:00:00                Iter   3   54.05%  Refine biparts1676 MB(24%)00:00:00                Iter   3   56.76%  Refine biparts1676 MB(24%)00:00:00                Iter   3   59.46%  Refine biparts1676 MB(24%)00:00:00                Iter   3   62.16%  Refine biparts1676 MB(24%)00:00:00                Iter   3   64.86%  Refine biparts1676 MB(24%)00:00:00                Iter   3   67.57%  Refine biparts1676 MB(24%)00:00:00                Iter   3   70.27%  Refine biparts1676 MB(24%)00:00:00                Iter   3   72.97%  Refine biparts1676 MB(24%)00:00:00                Iter   3   75.68%  Refine biparts1676 MB(24%)00:00:00                Iter   3   78.38%  Refine biparts1676 MB(24%)00:00:00                Iter   3   81.08%  Refine biparts1676 MB(24%)00:00:00                Iter   3   83.78%  Refine biparts1676 MB(24%)00:00:00                Iter   3   86.49%  Refine biparts1676 MB(24%)00:00:00                Iter   3   89.19%  Refine biparts1676 MB(24%)00:00:00                Iter   3   91.89%  Refine biparts1676 MB(24%)00:00:00                Iter   3   94.59%  Refine biparts1676 MB(24%)00:00:00                Iter   3   97.30%  Refine biparts1676 MB(24%)00:00:00                Iter   3  100.00%  Refine biparts1676 MB(24%)00:00:00                Iter   3  102.70%  Refine biparts1676 MB(24%)00:00:00                Iter   3  100.00%  Refine biparts</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1676 MB(24%)00:00:00                Iter   4    5.41%  Refine biparts1676 MB(24%)00:00:00                Iter   4    8.11%  Refine biparts1676 MB(24%)00:00:00                Iter   4   10.81%  Refine biparts1676 MB(24%)00:00:00                Iter   4   13.51%  Refine biparts1676 MB(24%)00:00:00                Iter   4   16.22%  Refine biparts1676 MB(24%)00:00:00                Iter   4   18.92%  Refine biparts1676 MB(24%)00:00:00                Iter   4   21.62%  Refine biparts1676 MB(24%)00:00:00                Iter   4   24.32%  Refine biparts1676 MB(24%)00:00:00                Iter   4   27.03%  Refine biparts1676 MB(24%)00:00:00                Iter   4   29.73%  Refine biparts1676 MB(24%)00:00:00                Iter   4   32.43%  Refine biparts1676 MB(24%)00:00:00                Iter   4   35.14%  Refine biparts1676 MB(24%)00:00:00                Iter   4   37.84%  Refine biparts1676 MB(24%)00:00:00                Iter   4   40.54%  Refine biparts1676 MB(24%)00:00:00                Iter   4   43.24%  Refine biparts1676 MB(24%)00:00:00                Iter   4   45.95%  Refine biparts1676 MB(24%)00:00:00                Iter   4   48.65%  Refine biparts1676 MB(24%)00:00:00                Iter   4   51.35%  Refine biparts1676 MB(24%)00:00:00                Iter   4   54.05%  Refine biparts1676 MB(24%)00:00:00                Iter   4   56.76%  Refine biparts1676 MB(24%)00:00:00                Iter   4   59.46%  Refine biparts1676 MB(24%)00:00:00                Iter   4   62.16%  Refine biparts1676 MB(24%)00:00:00                Iter   4   64.86%  Refine biparts1676 MB(24%)00:00:00                Iter   4   67.57%  Refine biparts1676 MB(24%)00:00:00                Iter   4   70.27%  Refine biparts1676 MB(24%)00:00:00                Iter   4   72.97%  Refine biparts1676 MB(24%)00:00:00                Iter   4   75.68%  Refine biparts1676 MB(24%)00:00:00                Iter   4   78.38%  Refine biparts1676 MB(24%)00:00:00                Iter   4   81.08%  Refine biparts1676 MB(24%)00:00:00                Iter   4   83.78%  Refine biparts1676 MB(24%)00:00:00                Iter   4   86.49%  Refine biparts1676 MB(24%)00:00:00                Iter   4   89.19%  Refine biparts1676 MB(24%)00:00:00                Iter   4   91.89%  Refine biparts1676 MB(24%)00:00:00                Iter   4   94.59%  Refine biparts1676 MB(24%)00:00:00                Iter   4   97.30%  Refine biparts1676 MB(24%)00:00:00                Iter   4  100.00%  Refine biparts1676 MB(24%)00:00:00                Iter   4  102.70%  Refine biparts1676 MB(24%)00:00:00                Iter   4  100.00%  Refine biparts</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1676 MB(24%)00:00:00                Iter   5    5.41%  Refine biparts1676 MB(24%)00:00:00                Iter   5    8.11%  Refine biparts1676 MB(24%)00:00:01                Iter   5   10.81%  Refine biparts1676 MB(24%)00:00:01                Iter   5  100.00%  Refine biparts</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1676 MB(24%)00:00:01                Iter   5  100.00%  Refine biparts</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>position_stack requires non-overlapping x intervals</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>position_stack requires non-overlapping x intervals</span>
<span class="r-plt img"><img src="perform_msa-1.png" alt="" width="700" height="433"></span>
<span class="r-plt img"><img src="perform_msa-2.png" alt="" width="700" height="433"></span>
<span class="r-in"></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Lucrezia Patruno, Daniele Ramazzotti, Dawid Nowak.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 1.9000.9000.9000.</p>
</div>

      </footer></div>

  


  

  </body></html>

