<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>perform_msa — perform_msa • REvoBC</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="perform_msa — perform_msa" />
<meta property="og:description" content="This function performs multi-sequence alignment and outputs statistics about mutations. 
Multiple Sequence Alignment is performed through the MUSCLE algorithm." />






<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  

  

  </head>

  <body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">REvoBC</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.99.6</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/1_introduction.html">Introduction</a>
    </li>
    <li>
      <a href="../articles/2_running_REvoBC.html">Running REvoBC</a>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/Nowak-Lab/REvoBC/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>perform_msa</h1>
    <small class="dont-index">Source: <a href='https://github.com/Nowak-Lab/REvoBC/blob/master/R/user_functions.R'><code>R/user_functions.R</code></a></small>
    <div class="hidden name"><code>perform_msa.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>This function performs multi-sequence alignment and outputs statistics about mutations. 
Multiple Sequence Alignment is performed through the <a href='https://bioconductor.org/packages/release/bioc/html/muscle.html'>MUSCLE</a> algorithm.</p>
    </div>

    <div class="ref-usage sourceCode"><pre class='sourceCode r'><code><span class='fu'>perform_msa</span><span class='op'>(</span><span class='va'>REvoBC_object</span>, <span class='va'>...</span><span class='op'>)</span></code></pre></div>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>REvoBC_object</th>
      <td><p>REvoBC object on which we want to perform msa.</p></td>
    </tr>
    <tr>
      <th>...</th>
      <td><p>Optional parameters (options and flags) passed to MUSCLE. See the <a href='http://www.drive5.com/muscle/muscle_userguide3.8.html'>original guide</a> for detailed information on all possible options.</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>REvoBC object with a new field named <code>alignment</code>, which is a list with the following fileds:</p><ul>
<li><p><code>msa_stringset</code>: output of MSA peformed with MUSCLE.</p></li>
<li><p><code>mutations_df</code>: tibble where each line corresponds to a position in a ASV, and the columns encode the following information:</p><ul>
<li><p>asv_names: name of the ASV</p></li>
<li><p>day_organ: sample</p></li>
<li><p>position_bc260: position of the alteration in the original barcode. Note that insertions
are assigned to the position that coincides with their beginning.</p></li>
<li><p>alt: type of alteration. wt = Wild Type (i.e. non-mutated position). sub = substitution. del = deletion. ins = insertion.</p></li>
<li><p>perc_in_sample: out of all sequences that map to a sample, the percentage of them that display the alteration.</p></li>
</ul></li>
</ul><p>. 
In addition, the following files are saved in the sub-folder "msa" created inthe output directory chosen by the user:</p><ul>
<li><p>dnastringset.fa: fasta file where the sequences are stored.</p></li>
<li><p>dnastringset.fa: same as fasta, but in csv format.</p></li>
<li><p>dnastringset_muscle-muscle_msa.fasta: fasta with the stringset resulted from MUSCLE.</p></li>
<li><p>ASV_alterationType_frequency.csv: number of alterations for each type in each ASV</p></li>
<li><p>mutations_frequency.csv: per sample normalized frequency of each alteration type, computed for each position of the barcode in each sample.</p></li>
<li><p>mutations_df.csv: content of mutations_df variable explained above.</p></li>
</ul>


    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <div class="ref-examples sourceCode"><pre class='sourceCode r'><code><span class='r-in'><span class='fu'><a href='https://rdrr.io/r/utils/data.html'>data</a></span><span class='op'>(</span><span class='va'>revo_analyzed</span><span class='op'>)</span></span>
<span class='r-in'><span class='va'>output_dir</span> <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/system.file.html'>system.file</a></span><span class='op'>(</span><span class='st'>"extdata"</span>, <span class='st'>"output"</span>, package <span class='op'>=</span> <span class='st'>"REvoBC"</span><span class='op'>)</span></span>
<span class='r-in'><span class='va'>revo_analyzed</span><span class='op'>$</span><span class='va'>output_directory</span> <span class='op'>=</span> <span class='va'>output_dir</span></span>
<span class='r-in'><span class='va'>revo_msa</span> <span class='op'>=</span> <span class='fu'>perform_msa</span><span class='op'>(</span><span class='va'>revo_analyzed</span><span class='op'>)</span></span>
<span class='r-wrn co'><span class='r-pr'>#&gt;</span> <span class='warning'>Warning: </span>attempt to set 'col.names' ignored</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span> </span>
<span class='r-out co'><span class='r-pr'>#&gt;</span> MUSCLE v3.8.31 by Robert C. Edgar</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span> </span>
<span class='r-out co'><span class='r-pr'>#&gt;</span> http://www.drive5.com/muscle</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span> This software is donated to the public domain.</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span> Please cite: Edgar, R.C. Nucleic Acids Res 32(5), 1792-97.</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span> </span>
<span class='r-out co'><span class='r-pr'>#&gt;</span> filee3133eaafc5f 21 seqs, max length 260, avg  length 175</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span> 1625 MB(23%)00:00:00                Iter   1    0.43%  K-mer dist pass 11625 MB(23%)00:00:00                Iter   1  100.00%  K-mer dist pass 1</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span> 1625 MB(23%)00:00:00                Iter   1    0.43%  K-mer dist pass 21625 MB(23%)00:00:00                Iter   1  100.00%  K-mer dist pass 2</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span> 1625 MB(23%)00:00:00                Iter   1    5.00%  Align node       1625 MB(23%)00:00:00                Iter   1   10.00%  Align node1625 MB(23%)00:00:00                Iter   1   15.00%  Align node1625 MB(23%)00:00:00                Iter   1   20.00%  Align node1625 MB(23%)00:00:00                Iter   1   25.00%  Align node1625 MB(23%)00:00:00                Iter   1   30.00%  Align node1625 MB(23%)00:00:00                Iter   1   35.00%  Align node1625 MB(23%)00:00:00                Iter   1   40.00%  Align node1625 MB(23%)00:00:00                Iter   1   45.00%  Align node1625 MB(23%)00:00:00                Iter   1   50.00%  Align node1625 MB(23%)00:00:00                Iter   1   55.00%  Align node1625 MB(23%)00:00:00                Iter   1   60.00%  Align node1625 MB(23%)00:00:00                Iter   1   65.00%  Align node1625 MB(23%)00:00:00                Iter   1   70.00%  Align node1625 MB(23%)00:00:00                Iter   1   75.00%  Align node1625 MB(23%)00:00:00                Iter   1   80.00%  Align node1625 MB(23%)00:00:00                Iter   1   85.00%  Align node1625 MB(23%)00:00:00                Iter   1   90.00%  Align node1625 MB(23%)00:00:00                Iter   1   95.00%  Align node1625 MB(23%)00:00:00                Iter   1  100.00%  Align node1625 MB(23%)00:00:00                Iter   1  100.00%  Align node</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span> 1625 MB(23%)00:00:00                Iter   1    4.76%  Root alignment1625 MB(23%)00:00:00                Iter   1    9.52%  Root alignment1625 MB(23%)00:00:00                Iter   1   14.29%  Root alignment1625 MB(23%)00:00:00                Iter   1   19.05%  Root alignment1625 MB(23%)00:00:00                Iter   1   23.81%  Root alignment1625 MB(23%)00:00:00                Iter   1   28.57%  Root alignment1625 MB(23%)00:00:00                Iter   1   33.33%  Root alignment1625 MB(23%)00:00:00                Iter   1   38.10%  Root alignment1625 MB(23%)00:00:00                Iter   1   42.86%  Root alignment1625 MB(23%)00:00:00                Iter   1   47.62%  Root alignment1625 MB(23%)00:00:00                Iter   1   52.38%  Root alignment1625 MB(23%)00:00:00                Iter   1   57.14%  Root alignment1625 MB(23%)00:00:00                Iter   1   61.90%  Root alignment1625 MB(23%)00:00:00                Iter   1   66.67%  Root alignment1625 MB(23%)00:00:00                Iter   1   71.43%  Root alignment1625 MB(23%)00:00:00                Iter   1   76.19%  Root alignment1625 MB(23%)00:00:00                Iter   1   80.95%  Root alignment1625 MB(23%)00:00:00                Iter   1   85.71%  Root alignment1625 MB(23%)00:00:00                Iter   1   90.48%  Root alignment1625 MB(23%)00:00:00                Iter   1   95.24%  Root alignment1625 MB(23%)00:00:00                Iter   1  100.00%  Root alignment1625 MB(23%)00:00:00                Iter   1  100.00%  Root alignment</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span> 1625 MB(23%)00:00:00                Iter   2    5.26%  Refine tree   1625 MB(23%)00:00:00                Iter   2   10.53%  Refine tree1625 MB(23%)00:00:00                Iter   2   15.79%  Refine tree1625 MB(23%)00:00:00                Iter   2   21.05%  Refine tree1625 MB(23%)00:00:00                Iter   2   26.32%  Refine tree1625 MB(23%)00:00:00                Iter   2   31.58%  Refine tree1625 MB(23%)00:00:00                Iter   2   36.84%  Refine tree1625 MB(23%)00:00:00                Iter   2   42.11%  Refine tree1625 MB(23%)00:00:00                Iter   2   47.37%  Refine tree1625 MB(23%)00:00:00                Iter   2   52.63%  Refine tree1625 MB(23%)00:00:00                Iter   2   57.89%  Refine tree1625 MB(23%)00:00:00                Iter   2   63.16%  Refine tree1625 MB(23%)00:00:00                Iter   2   68.42%  Refine tree1625 MB(23%)00:00:00                Iter   2   73.68%  Refine tree1625 MB(23%)00:00:00                Iter   2   78.95%  Refine tree1625 MB(23%)00:00:00                Iter   2   84.21%  Refine tree1625 MB(23%)00:00:00                Iter   2   89.47%  Refine tree1625 MB(23%)00:00:00                Iter   2   94.74%  Refine tree1625 MB(23%)00:00:00                Iter   2  100.00%  Refine tree1625 MB(23%)00:00:00                Iter   2  100.00%  Refine tree</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span> 1625 MB(23%)00:00:00                Iter   2    4.76%  Root alignment1625 MB(23%)00:00:00                Iter   2    9.52%  Root alignment1625 MB(23%)00:00:00                Iter   2   14.29%  Root alignment1625 MB(23%)00:00:00                Iter   2   19.05%  Root alignment1625 MB(23%)00:00:00                Iter   2   23.81%  Root alignment1625 MB(23%)00:00:00                Iter   2   28.57%  Root alignment1625 MB(23%)00:00:00                Iter   2   33.33%  Root alignment1625 MB(23%)00:00:00                Iter   2   38.10%  Root alignment1625 MB(23%)00:00:00                Iter   2   42.86%  Root alignment1625 MB(23%)00:00:00                Iter   2   47.62%  Root alignment1625 MB(23%)00:00:00                Iter   2   52.38%  Root alignment1625 MB(23%)00:00:00                Iter   2   57.14%  Root alignment1625 MB(23%)00:00:00                Iter   2   61.90%  Root alignment1625 MB(23%)00:00:00                Iter   2   66.67%  Root alignment1625 MB(23%)00:00:00                Iter   2   71.43%  Root alignment1625 MB(23%)00:00:00                Iter   2   76.19%  Root alignment1625 MB(23%)00:00:00                Iter   2   80.95%  Root alignment1625 MB(23%)00:00:00                Iter   2   85.71%  Root alignment1625 MB(23%)00:00:00                Iter   2   90.48%  Root alignment1625 MB(23%)00:00:00                Iter   2   95.24%  Root alignment1625 MB(23%)00:00:00                Iter   2  100.00%  Root alignment1625 MB(23%)00:00:00                Iter   2  100.00%  Root alignment</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span> 1625 MB(23%)00:00:00                Iter   2  100.00%  Root alignment</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span> 1625 MB(23%)00:00:00                Iter   3    5.13%  Refine biparts1625 MB(23%)00:00:00                Iter   3    7.69%  Refine biparts1625 MB(23%)00:00:00                Iter   3   10.26%  Refine biparts1625 MB(23%)00:00:00                Iter   3   12.82%  Refine biparts1625 MB(23%)00:00:00                Iter   3   15.38%  Refine biparts1625 MB(23%)00:00:00                Iter   3   17.95%  Refine biparts1625 MB(23%)00:00:00                Iter   3   20.51%  Refine biparts1625 MB(23%)00:00:00                Iter   3   23.08%  Refine biparts1625 MB(23%)00:00:00                Iter   3   25.64%  Refine biparts1625 MB(23%)00:00:00                Iter   3   28.21%  Refine biparts1625 MB(23%)00:00:00                Iter   3   30.77%  Refine biparts1625 MB(23%)00:00:00                Iter   3   33.33%  Refine biparts1625 MB(23%)00:00:00                Iter   3   35.90%  Refine biparts1625 MB(23%)00:00:00                Iter   3   38.46%  Refine biparts1625 MB(23%)00:00:00                Iter   3   41.03%  Refine biparts1625 MB(23%)00:00:00                Iter   3   43.59%  Refine biparts1625 MB(23%)00:00:00                Iter   3   46.15%  Refine biparts1625 MB(23%)00:00:00                Iter   3   48.72%  Refine biparts1625 MB(23%)00:00:00                Iter   3   51.28%  Refine biparts1625 MB(23%)00:00:00                Iter   3   53.85%  Refine biparts1625 MB(23%)00:00:00                Iter   3   56.41%  Refine biparts1625 MB(23%)00:00:00                Iter   3   58.97%  Refine biparts1625 MB(23%)00:00:00                Iter   3   61.54%  Refine biparts1625 MB(23%)00:00:00                Iter   3   64.10%  Refine biparts1625 MB(23%)00:00:00                Iter   3   66.67%  Refine biparts1625 MB(23%)00:00:00                Iter   3   69.23%  Refine biparts1625 MB(23%)00:00:00                Iter   3   71.79%  Refine biparts1625 MB(23%)00:00:00                Iter   3   74.36%  Refine biparts1625 MB(23%)00:00:00                Iter   3   76.92%  Refine biparts1625 MB(23%)00:00:00                Iter   3   79.49%  Refine biparts1625 MB(23%)00:00:00                Iter   3   82.05%  Refine biparts1625 MB(23%)00:00:00                Iter   3   84.62%  Refine biparts1625 MB(23%)00:00:00                Iter   3   87.18%  Refine biparts1625 MB(23%)00:00:00                Iter   3   89.74%  Refine biparts1625 MB(23%)00:00:00                Iter   3   92.31%  Refine biparts1625 MB(23%)00:00:00                Iter   3   94.87%  Refine biparts1625 MB(23%)00:00:00                Iter   3   97.44%  Refine biparts1625 MB(23%)00:00:00                Iter   3  100.00%  Refine biparts1625 MB(23%)00:00:00                Iter   3  102.56%  Refine biparts1625 MB(23%)00:00:00                Iter   3  100.00%  Refine biparts</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span> 1625 MB(23%)00:00:00                Iter   4    5.13%  Refine biparts1625 MB(23%)00:00:00                Iter   4    7.69%  Refine biparts1625 MB(23%)00:00:00                Iter   4   10.26%  Refine biparts1625 MB(23%)00:00:00                Iter   4   12.82%  Refine biparts1625 MB(23%)00:00:00                Iter   4   15.38%  Refine biparts1625 MB(23%)00:00:00                Iter   4   17.95%  Refine biparts1625 MB(23%)00:00:00                Iter   4   20.51%  Refine biparts1625 MB(23%)00:00:00                Iter   4   23.08%  Refine biparts1625 MB(23%)00:00:00                Iter   4   25.64%  Refine biparts1625 MB(23%)00:00:00                Iter   4   28.21%  Refine biparts1625 MB(23%)00:00:00                Iter   4   30.77%  Refine biparts1625 MB(23%)00:00:00                Iter   4   33.33%  Refine biparts1625 MB(23%)00:00:00                Iter   4   35.90%  Refine biparts1625 MB(23%)00:00:00                Iter   4   38.46%  Refine biparts1625 MB(23%)00:00:00                Iter   4   41.03%  Refine biparts1625 MB(23%)00:00:00                Iter   4   43.59%  Refine biparts1625 MB(23%)00:00:00                Iter   4   46.15%  Refine biparts1625 MB(23%)00:00:00                Iter   4   48.72%  Refine biparts1625 MB(23%)00:00:00                Iter   4   51.28%  Refine biparts1625 MB(23%)00:00:00                Iter   4   53.85%  Refine biparts1625 MB(23%)00:00:00                Iter   4   56.41%  Refine biparts1625 MB(23%)00:00:00                Iter   4   58.97%  Refine biparts1625 MB(23%)00:00:00                Iter   4   61.54%  Refine biparts1625 MB(23%)00:00:00                Iter   4   64.10%  Refine biparts1625 MB(23%)00:00:00                Iter   4   66.67%  Refine biparts1625 MB(23%)00:00:00                Iter   4   69.23%  Refine biparts1625 MB(23%)00:00:00                Iter   4   71.79%  Refine biparts1625 MB(23%)00:00:00                Iter   4   74.36%  Refine biparts1625 MB(23%)00:00:00                Iter   4   76.92%  Refine biparts1625 MB(23%)00:00:00                Iter   4   79.49%  Refine biparts1625 MB(23%)00:00:00                Iter   4   82.05%  Refine biparts1625 MB(23%)00:00:00                Iter   4   84.62%  Refine biparts1625 MB(23%)00:00:00                Iter   4   87.18%  Refine biparts1625 MB(23%)00:00:00                Iter   4   89.74%  Refine biparts1625 MB(23%)00:00:00                Iter   4   92.31%  Refine biparts1625 MB(23%)00:00:00                Iter   4   94.87%  Refine biparts1625 MB(23%)00:00:00                Iter   4   97.44%  Refine biparts1625 MB(23%)00:00:00                Iter   4  100.00%  Refine biparts1625 MB(23%)00:00:00                Iter   4  102.56%  Refine biparts1625 MB(23%)00:00:00                Iter   4  100.00%  Refine biparts</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span> 1625 MB(23%)00:00:00                Iter   5    5.13%  Refine biparts1625 MB(23%)00:00:00                Iter   5    7.69%  Refine biparts1625 MB(23%)00:00:00                Iter   5   10.26%  Refine biparts1625 MB(23%)00:00:00                Iter   5   12.82%  Refine biparts1625 MB(23%)00:00:00                Iter   5   15.38%  Refine biparts1625 MB(23%)00:00:00                Iter   5   17.95%  Refine biparts1625 MB(23%)00:00:00                Iter   5  100.00%  Refine biparts</span>
<span class='r-out co'><span class='r-pr'>#&gt;</span> 1625 MB(23%)00:00:00                Iter   5  100.00%  Refine biparts</span>
<span class='r-wrn co'><span class='r-pr'>#&gt;</span> <span class='warning'>Warning: </span>position_stack requires non-overlapping x intervals</span>
<span class='r-wrn co'><span class='r-pr'>#&gt;</span> <span class='warning'>Warning: </span>position_stack requires non-overlapping x intervals</span>
<span class='r-in'></span>
</code></pre></div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p><p>Developed by Lucrezia Patruno, Daniele Ramazzotti, Dawid Nowak.</p></p>
</div>

<div class="pkgdown">
  <p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 1.6.1.9001.</p></p>
</div>

      </footer>
   </div>

  


  

  </body>
</html>


