<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Running EvoTraceR • EvoTraceR</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Running EvoTraceR">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">EvoTraceR</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.0.3</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/1_introduction.html">Introduction</a></li>
    <li><a class="dropdown-item" href="../articles/2_running_REvoBC.html">Running EvoTraceR</a></li>
  </ul>
</li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/Nowak-Lab/EvoTraceR/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Running EvoTraceR</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/Nowak-Lab/EvoTraceR/blob/master/vignettes/2_running_REvoBC.Rmd" class="external-link"><code>vignettes/2_running_REvoBC.Rmd</code></a></small>
      <div class="d-none name"><code>2_running_REvoBC.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="setup-the-environement">Setup the environement<a class="anchor" aria-label="anchor" href="#setup-the-environement"></a>
</h2>
<p><em>EvoTraceR</em> is the computational tool inside the project
EvoBC. After performing EvoBC experiments in the lab, this tool can be
exploited to perform the analysis of Amplicon Sequence Variants. In
order to perform the analysis, there are multiple functions that perform
the different steps of the analysis pipeline. The first step initializes
the EvoTraceR object, which is a list that gets populated step-by-step
with multiple fields containing the result of the analysis.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://github.com/Nowak-Lab/EvoTraceR" class="external-link">"EvoTraceR"</a></span><span class="op">)</span></span>
<span><span class="co"># Run in case reticulate was setup manually:</span></span>
<span><span class="co"># library(reticulate)</span></span>
<span><span class="co"># reticulate::use_condaenv('EvoTraceR_enviroment')</span></span></code></pre></div>
<p>The initialization function computes the set of Amplicon Sequence
Variants. It calls trimmomatic and flash tools to perform adapters
trimming, discard low quality bases and merge forward and reverse reads.
The main parameters for the initialization are the directory where the
output will be stored and one of the following: input directory: in this
case users have the fastq files and want to perform the whole analysis
using EvoTraceR. In this case they need to provide the path to the
folder containing the fastq (which may be compressed), and then they can
provide any additional parameter for dada2 functions. trimmomatic_path:
Local path to the executable of <em>Trimmomatic</em>. flash_path. Local
path to the executable of <em>Flash</em>. Important: each fastq file
pair (forward and reverse) must be associated to the sample on which the
experiment was performed. The sample can be either the organ of the
mouse or an ID for the timepoint in case longitudinal data is provided.
In order for EvoTraceR to associate each fastq pair to its sample users
have two options: one is to name the fastq following this pattern:
FILEPREFIX_SAMPLE_FILESUFFIX_RX.fastq, where FILEPREFIX and FILESUFFIX
can be any string (without the “_” character). In this case EvoTraceR
will consider as sample ID the second field separated by the underscore.
The second option is to provide a list that associates each filename
(without the suffix _R1 and _R2) to the corresponding sample. For
example, if we have forward and reverse files named file1_R1.fastq and
file1_R2.fastq that correspond to organ PRL (code for Prostate Left),
than the parameter should be set as: map_file_sample = c(“file1” =
“PRL”)).</p>
<p>Users can skip the initialization step for this vignette, as the
output has been pre-computed and can be loaded by calling
data(EvoTraceR_initialized).</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">input_dir</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"input"</span>, package <span class="op">=</span> <span class="st">"EvoTraceR"</span><span class="op">)</span></span>
<span><span class="va">output_dir</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"output"</span>, package <span class="op">=</span> <span class="st">"EvoTraceR"</span><span class="op">)</span></span></code></pre></div>
<p>We can now perform the initialization.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">EvoTraceR_object</span> <span class="op">=</span> <span class="fu"><a href="../reference/initialize_EvoTraceR.html">initialize_EvoTraceR</a></span><span class="op">(</span>input_dir <span class="op">=</span> <span class="va">input_dir</span>, </span>
<span>                                        output_dir <span class="op">=</span> <span class="va">output_dir</span>, </span>
<span>                                        trimmomatic_path <span class="op">=</span> <span class="st">'~/Trimmomatic 0.xx/trimmomatic-0.xx.jar'</span>,</span>
<span>                                        flash_path <span class="op">=</span> <span class="st">'~/*/FLASH-x.x.x/flash'</span><span class="op">)</span></span></code></pre></div>
<p>During the above computation, all unique sequences from each sample
have been identified. In the next steps these sequences will be
cleaned.</p>
</div>
<div class="section level2">
<h2 id="analysis-and-pairwise-alignment-of-asvs">Analysis and pairwise alignment of ASVs<a class="anchor" aria-label="anchor" href="#analysis-and-pairwise-alignment-of-asvs"></a>
</h2>
<p>This function performs the analysis on ASV sequences identified by
the previous steps and aligns them to the reference sequence. First, it
pools together those sequences characterized by a Hamming distance
equals or lower than 2. Then it performs pairwise alignment using
Needleman-Wunsch global alignment algorithm implemented in function
<em>pairwiseAlignment</em> in package <em>pwalign</em>, aligning each
sequence to the original barcode considered in the analysis (See the
pwalign documentation for more details). After identifying all indels
(insertions are identified by their start position and number of
nucleoides inserted, while deletions are identified by their start and
end position), it removes the ones that are too small and don’t span any
cut site: to perform this filter, it expands the start and end position
by the number of bases specified in the parameter
<em>cleaning_window</em>, it counts the number of cut sites spanned
after the expansion and it removes those that span 0 sites. Next, it
pools together those sequences that differ by one another only by
substitutions. Then it computes the normalized counts of each ASV in
each sample, dividing each count by the total number of sequences for
each sample and multiplying by 1e6, yelding Counts Per Million (CPM).
After normalization, counts are summed in each ASV, and those sequences
showing a frequency lower than the threshold specified in parameter
<em>asv_count_cutoff</em> are removed. Finally, it performs filtering of
the sequences based on their flanking sequences. Then, using the CPM it
computes different statistics for the final ASVS, storing: the relative
frequency of all ASVs in each sample. the relative frequency of each ASV
in the samples. the counts for each ASV normalized to the counts of the
sample with maximum frequency. the frequency of the different ASVs in
each sample.</p>
<p>Users can skip this step for this vignette, as the output has been
pre-computed and can be loaded by calling data(EvoTraceR_analyzed).</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">EvoTraceR_object</span> <span class="op">=</span> <span class="fu"><a href="../reference/asv_analysis.html">asv_analysis</a></span><span class="op">(</span>EvoTraceR_object <span class="op">=</span> <span class="va">EvoTraceR_object</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">EvoTraceR_object</span><span class="op">)</span></span></code></pre></div>
<p>There are new fields: clean_asv_dataframe which contains the counts
detected for each ASV after filtering (i.e. after filtering for
contamination identification of NMBC). reference is the information on
the reference sequence considered for the analysis. clean_asv_dataframe:
ASV sequences identified post-filtering (contamination removed,
sequences with a similarity higher than <em>pid_cutoff_nmbc</em> to the
original barcode aggregated to it and ASVs named in increasing order
(ASV01, ASV02, etc.) according to their total counts. reference: Info
about the reference sequence used for the current analysis. statistics:
is a new list that contains all the results of the computation (see the
manual for a detailed explanation). Among its fields there are
asv_df_percentages, which contains the normalized counts, and
asv_diversity_perSample which contains the measures of clonal
heterogeneity. alignment: is a new list containing the result of the
alignment step.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">EvoTraceR_object</span> <span class="op">=</span> <span class="fu"><a href="../reference/analyse_mutations.html">analyse_mutations</a></span><span class="op">(</span><span class="va">EvoTraceR_object</span><span class="op">)</span></span></code></pre></div>
<p>After alignment, EvoTraceR identifies, for each position in the
original reference, the number of alterations found on the different
ASVs. This computation stores the result in a new field
mutations_df.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">EvoTraceR_object</span> <span class="op">=</span> <span class="fu"><a href="../reference/infer_phylogeny.html">infer_phylogeny</a></span><span class="op">(</span><span class="va">EvoTraceR_object</span><span class="op">)</span></span></code></pre></div>
<p>This function reconstructs the phylogenetic tree using the
Cassiopeia’s suite. It uses the greedy algorithm, that iteratively
splits the set of ASVs in clusters based on the most common mutation at
each iteration.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">EvoTraceR_object</span> <span class="op">=</span> <span class="fu"><a href="../reference/create_df_summary.html">create_df_summary</a></span><span class="op">(</span><span class="va">EvoTraceR_object</span><span class="op">)</span></span></code></pre></div>
<p>Finally, this function combines in one dataframe all the data that
have been computed for the phylogenetic analysis.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/save.html" class="external-link">save</a></span><span class="op">(</span><span class="va">EvoTraceR_object</span>, file<span class="op">=</span><span class="st">"./data/EvoTraceR_object.RData"</span><span class="op">)</span></span></code></pre></div>
<p>Save EvoTraceR_object</p>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Lucrezia Patruno, Daniele Ramazzotti, Armin Scheben, Stephen Staklinski, Rebecca Hassett, Dylan Riffle, Dawid Nowak.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.9000.</p>
</div>

    </footer>
</div>





  </body>
</html>
